feat: Add a SessionOwner ObjectStatus and allow InScenePlaced network objects to be distributed

EmandM opened the pull request:
<!-- Replace this block with what this PR does and why. Describe what you'd like reviewers to know, how you applied the engineering principles, and any interesting tradeoffs made. Delete bullet points below that don't apply, and update the changelog section as appropriate. -->

During the development of NGO DA it was surfaced that there are some network objects that we don't want to distribute when a new client joins. The workaround at the time was to not distribute any in-scene placed network objects when a new client joins.

This PR fixes this issue via adding a `NetworkObject.OwnershipStatus.SessionOwner` to mark objects that should be distributable, but should only ever belong to the current session owner. It also now allows in-scene placed network objects to be distributed.

The `NetworkObject.OwnershipStatus.SessionOwner` flag will override all other `OwnershipStatus` flags, and cannot be set in conjunction with any other flag.

<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->

<!-- Add RFC link here if applicable. -->

## Changelog

- Added: `NetworkObject.OwnershipStatus.SessionOwner` flag. This flag will override all other  `OwnershipStatus` flags.
- Changed: In-scene placed network objects can now be distributed when a new client joins.

## Testing and Documentation

- Includes unit tests.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

EmandM committed:
Initial pass on SessionOwner ownership flag

EmandM committed:
feat: Add SessionOwner OwnershipStatus flag

NoelStephensUnity committed:
Merge branch 'feat/session_owner_only_network_object_distribution' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into feat/session_owner_only_network_object_distribution

NoelStephensUnity committed:
update

removing additional session owner accessor.

EmandM committed:
Add OwnershipPermissions tests

EmandM committed:
fix client connect

EmandM committed:
Revert "fix client connect"

This reverts commit 3c3b35444200da308dd07e1dd2e7b591fa0842d5.

NoelStephensUnity committed:
update

object distribution for in-scene placed NetworkObjects needs to use the InScenePlacedSourceGlobalObjectIdHash when building an object distribution list.

NoelStephensUnity committed:
Merge branch 'feat/session_owner_only_network_object_distribution' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into feat/session_owner_only_network_object_distribution

EmandM committed:
Add changelog

EmandM committed:
Merge branch 'feat/session_owner_only_network_object_distribution' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into feat/session_owner_only_network_object_distribution

EmandM committed:
Remove unnecessary change

EmandM committed:
Remove Settings.json

EmandM committed:
Reword CHANGELOG

NoelStephensUnity committed:
fix

DAHost should not distribute session owner permission NetworkObjects.
When client is promoted to session owner, for now newly promoted client takes ownership of NetworkObjects that have the SessionOwner permission set.
Only prevent non-session owners from taking ownership of a NetworkObject with the SessionOwner permission set.

NoelStephensUnity committed:
Merge branch 'feat/session_owner_only_network_object_distribution' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into feat/session_owner_only_network_object_distribution

NoelStephensUnity committed:
test fix

Avoid the RemoveOwnership client-server only method.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/session_owner_only_network_object_distribution

NoelStephensUnity committed:
style

Visual studio code cleanup likes to sort by alpha... fixing for our standards.

NoelStephensUnity reviewed (approved):
Awesome!
:godmode:

NoelStephensUnity committed:
update

Adding check for session owner trying to change ownership to a non-session owner client.

NoelStephensUnity committed:
test

Adding an additional validation that a non-session owner cannot change ownership and that a session owner cannot change ownership to a non-session owner when the NetworkObject in question has the SessionOwner permissions set.

jabbacakes commented:
This seems like it might need docs updates? 👀
